drivers/adc/adc_esp32_dma.c

```diff
@@ -0,0 +1,484 @@
+
+#endif /* SOC_GDMA_SUPPORTED */
+
+#define ADC_DMA_BUFFER_SIZE DMA_DESCRIPTOR_BUFFER_MAX_SIZE_4B_ALIGNED
+
+#if SOC_GDMA_SUPPORTED
+
+static void IRAM_ATTR adc_esp32_dma_conv_done(const struct device *dma_dev, void *user_data,
+					      uint32_t channel, int status)
+{
+	ARG_UNUSED(dma_dev);
+	ARG_UNUSED(status);
+
+	const struct device *dev = user_data;
+	struct adc_esp32_data *data = dev->data;
+
+	k_sem_give(&data->dma_conv_wait_lock);
+}
+
+static int adc_esp32_dma_start(const struct device *dev, uint8_t *buf, size_t len)
+{
+	const struct adc_esp32_conf *conf = dev->config;
+	int err = 0;
+
+	struct dma_config dma_cfg = {0};
+	struct dma_status dma_status = {0};
+	struct dma_block_config dma_blk = {0};
+
+	err = dma_get_status(conf->dma_dev, conf->dma_channel, &dma_status);
+	if (err) {
+		LOG_ERR("Unable to get dma channel[%u] status (%d)",
+			(unsigned int)conf->dma_channel, err);
+		return -EINVAL;
+	}
+
+	if (dma_status.busy) {
+		LOG_ERR("dma channel[%u] is busy!", (unsigned int)conf->dma_channel);
+		return -EBUSY;
+	}
+
+	unsigned int key = irq_lock();
+
+	dma_cfg.channel_direction = PERIPHERAL_TO_MEMORY;
+	dma_cfg.dma_callback = adc_esp32_dma_conv_done;
+	dma_cfg.user_data = (void *)dev;
+	dma_cfg.dma_slot = ESP_GDMA_TRIG_PERIPH_ADC0;
+	dma_cfg.block_count = 1;
+	dma_cfg.head_block = &dma_blk;
+	dma_blk.block_size = len;
+	dma_blk.dest_address = (uint32_t)buf;
+
+	err = dma_config(conf->dma_dev, conf->dma_channel, &dma_cfg);
+	if (err) {
+		LOG_ERR("Error configuring dma (%d)", err);
+		goto unlock;
+	}
+
+	err = dma_start(conf->dma_dev, conf->dma_channel);
+	if (err) {
+		LOG_ERR("Error starting dma (%d)", err);
+		goto unlock;
+	}
+
+unlock:
+	irq_unlock(key);
+	return err;
+}
+
+static int adc_esp32_dma_stop(const struct device *dev)
+{
+	const struct adc_esp32_conf *conf = dev->config;
+	unsigned int key = irq_lock();
+	int err = 0;
+
+	err = dma_stop(conf->dma_dev, conf->dma_channel);
+	if (err) {
+		LOG_ERR("Error stopping dma (%d)", err);
+	}
+
+	irq_unlock(key);
+	return err;
+}
+
+#else
+
+static IRAM_ATTR void adc_esp32_dma_intr_handler(void *arg)
+{
+	if (arg == NULL) {
+		return;
+	}
+
+	const struct device *dev = arg;
+	struct adc_esp32_data *data = dev->data;
+
+	bool conv_completed = adc_dma_check_event(ADC_DMA_DEV, ADC_DMA_INTR_MASK);
+
+	if (conv_completed) {
+		adc_dma_digi_clr_intr(ADC_DMA_DEV, ADC_DMA_INTR_MASK);
+
+		k_sem_give(&data->dma_conv_wait_lock);
+	}
+}
+
+#endif /* SOC_GDMA_SUPPORTED */
+
+static int adc_esp32_fill_digi_ctrlr_cfg(const struct device *dev, const struct adc_sequence *seq,
+					 uint32_t sample_freq_hz,
+					 adc_digi_pattern_config_t *pattern_config,
+					 adc_hal_digi_ctrlr_cfg_t *adc_hal_digi_ctrlr_cfg,
+					 uint32_t *pattern_len, uint32_t *unit_attenuation)
+{
+	const struct adc_esp32_conf *conf = dev->config;
+	struct adc_esp32_data *data = dev->data;
+
+	adc_digi_pattern_config_t *adc_digi_pattern_config = pattern_config;
+	const uint32_t unit_atten_uninit = 999;
```

Comment:
This number `999` sounds magic number, which is not. I suggested making this global with
`#define UNIT_ATTEN_UNINIT UINT32_MAX` for example.